src/utils/sanitize-hast.js

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2019 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+const _ = require('lodash');
+const sanitize = require('hast-util-sanitize');
+const GITHUB_SCHEMA = require('hast-util-sanitize/lib/github');
+
+// Define helix specific customizations on top of the default GitHub schema
+const HELIX_SCHEMA_CUSTOMIZATIONS = {
+  tagNames: ['esi:include'],
```

Comment:
Agreed. But then again, white-listing is the safest XSS approach.
The best security is always "deny-all" and open up what you need. I'd rather let the customer extend the schema with their own as needed, than just "allow-all".

What do you think?

Reply:
I don't know if this is still an issue with your new implementation. For normal HTML elements, let's use a whitelist of known good elements. For custom elements, let's assume that they are all good, because they would have to be inserted by the developer of the site.

As long as we don't have a good way of configuring the XSS filter (and we probably should not introduce a configuration anyway), we should have defaults that work, even when you are using extensions.

Reply:
It still would be. DOMPurify will remove any custom elements by default.
We'd have to follow their proposed solution to https://github.com/cure53/DOMPurify/issues/184 in:
https://github.com/cure53/DOMPurify/blob/master/test/test-suite.js#L294-L312

> we probably should not introduce a configuration anyway

I'm not sure about that one. We can have a catch-all solution for `<\w+-\w+` tags, but you'd still end up with edge cases that customers may need to have.
- What about `<customtag/>`?
- Should we support custom properties as well? `<custom-tag custom-prop/>`
- What about `<esi:custom/>`?

Assuming all custom elements are "good" is the same as saying all authors know what they are doing and their sessions cannot be compromised. While I'm ok with that argument for DOM Clobbering topic as DOMPurifies is applied after that anyway and secures the response, I'm really not convinved here.

My proposal would be to "white-list" all the `esi:` tags we need, but lock down the rest and let it be configured for each project by extending the default DOMPurify config (and abstracting it obviously as it's an implementation detail).

Just a random thought, but I would be happy as a front-end dev to define something like:
`.helix-config.json`
```js
{
  "allow-custom-html-tags": true, // allow all
  "allow-custom-html-tags": ["my-tag", "my-other-tag"], // allow some
  "allow-custom-html-attributes": true, // allow all
  "allow-custom-html-attributes": ["my-attribute", "my-other-attribute"], // allow some
}
```
i.e. the same I would define my linting configs for instance.

Reply:
> What about `<customtag/>`?
not a Custom Element, according to [spec](http://w3c.github.io/webcomponents/spec/custom/#prod-potentialcustomelementname): forbidden.
> Should we support custom properties as well? `<custom-tag custom-prop/>`
within a custom element, yes.
> What about `<esi:custom/>`?

Neither a proper Custom Element, nor part of the ESI language (https://www.w3.org/TR/esi-lang): forbidden

For ESI, we need a while list:
* `esi:include`
* `esi:remove`

For [Custom Elements](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements) we don't know what's coming in advance, so we can assume that the developer who is building the VDOM Matcher to produce `my-foo-element` is not trying to attack the other developer, working on the same code base, who is implementing `class MyFooElement extends HTMLParagraphElement…` on the client-side.

Reply:
I generally agree. But I'd really like to see some potential attack vectors described and tests for them being written. At the moment, I can't come up with any, but maybe @lkrapf or @asanso can?

Reply:
I'd only allow the ESI tags that are actually supported by Fastly (complete list above)